Implement carousel and add Fira Sans font.

thinking-tower opened the pull request:
Changes:
- Added a Carousel component.
- Added Fira Sans font.

#10 is implemented here.

thinking-tower committed:
Changes:
	- Added a Carousel component.
	- Added Fira Sans font.

thinking-tower committed:
Changes:
	- Added a Carousel component.
	- Added Fira Sans font.

thinking-tower committed:
Remove console.log. Add newline at the end of files.

thinking-tower committed:
Remove console.log. Add newline at the end of files.

thinking-tower committed:
Remove comments here since they're referenced in the issue.

thinking-tower committed:
Remove comments here since they're referenced in the issue.

thinking-tower commented:
## Preview
### Desktop
<img width="1421" alt="Screen Shot 2020-06-01 at 10 23 24 PM" src="https://user-images.githubusercontent.com/37529096/83400315-b770ff00-a456-11ea-8d57-f9ffcafde77f.png">

To make it overlapping, at the moment you'd have to set `itemsPerPage`. For example although we only have 6 items, by setting `itemsPerPage={20}` we get this:
<img width="1421" alt="Screen Shot 2020-06-01 at 10 27 35 PM" src="https://user-images.githubusercontent.com/37529096/83400533-1f274a00-a457-11ea-8287-ef9cff505832.png">

Sliding to the fifth index produces this:
<img width="1419" alt="Screen Shot 2020-06-01 at 10 28 26 PM" src="https://user-images.githubusercontent.com/37529096/83400638-50a01580-a457-11ea-9057-d02502f94314.png">

### Mobile
<img width="348" alt="Screen Shot 2020-06-01 at 10 29 38 PM" src="https://user-images.githubusercontent.com/37529096/83400703-68779980-a457-11ea-985c-db1d653d9dd8.png">

### Code
In Homepage.js,

```
import React from 'react';
import fbIcon from '../static/images/f_logo_RGB-Blue_58.png';
import twitterIcon from '../static/images/Twitter_Social_Icon_Circle_Color.png';
import '../static/css/homepage.css';

import Carousel from '../components/Carousel';

function Homepage() {
  return (
    <div>
      <header>
        <div className="header-text-container">
          <h1>The Template Project</h1>
          <p>A one stop shop for beautifully designed ReactJS templates</p>
        </div>
        <div className="social-icons-container">
          <a
            href="https://www.facebook.com/"
            target="_blank"
            rel="noopener noreferrer"
          >
            <img src={fbIcon} alt="fb icon" />
          </a>
          <a
            href="https://www.twitter.com/"
            target="_blank"
            rel="noopener noreferrer"
          >
            <img src={twitterIcon} alt="fb icon" />
          </a>
        </div>
      </header>
      <Carousel
        itemsPerPage={20}
      >
        {[...Array(6).keys()].map(i => ( 
           <img key={i} src={i%2==0 ? fbIcon : twitterIcon} />
        ))}
      </Carousel>
    </div>
  );
}


export default Homepage;
```
## Issues

- On mobile, I haven't managed to get the carousel buttons (blue in screenshots) to be centered on mobile yet

## TODO

-  Get assets for buttons
-  Project text should reflect whichever template is the current slide

## Note
#10 is implemented here.

## References
- How to build a carousel:
     - [How to code a carousel with HTML, CSS and JavaScript - from scratch (part 1)](https://www.youtube.com/watch?v=VYsVOamdB0g)
     - [How To Create An Image Slider In HTML, CSS & Javascript](https://www.youtube.com/watch?v=KcdBOoK3Pfw)
     - [brainhubeu/react-carousel](https://github.com/brainhubeu/react-carousel)
     - [metafizzy/flickity](https://github.com/metafizzy/flickity)
     - [kenwheeler/slick](https://github.com/kenwheeler/slick)
- Import font
     - [How to import Google Web Font in CSS file](https://stackoverflow.com/questions/14676613/how-to-import-google-web-font-in-css-file)

kipply commented on src/components/Carousel/Carousel.module.css:
```diff
@@ -0,0 +1,66 @@
+.carousel {
```
I think we can make one file in `src/components/Carousel.css`

kipply commented on src/index.css:
```diff
@@ -1,3 +1,5 @@
+@import url('https://fonts.googleapis.com/css2?family=Fira+Sans:wght@400;500&display=swap');
```
this is already imported in public.html

thinking-tower replied:
Will remove.

kipply commented on src/components/Carousel/index.js:
```diff
@@ -0,0 +1,3 @@
+import Carousel from './Carousel';
+
```
can we export carousel and import from the carousel file? Also it would be nice to add an example carousel into homepage.js (you may need to update your branch) 

Use images from https://unsplash.com/ for previews since it'll look more like what we're going for

thinking-tower replied:
Any images from unsplash.com?

kipply replied:
they're placeholders, doesn't matter.

kipply commented on src/components/Carousel/Carousel.module.css:
```diff
@@ -0,0 +1,66 @@
+    grid-column-start: 11;
+}
+
+.carousel__buttons-container button {
```
I think the double underscores in the names are verbose and not very useful. `carousel-buttons-container` should be sufficient.

thinking-tower replied:
So would you say for this I can just avoid using the BEM notation?

kipply replied:
yeah this isn't a particularly big site. Especially since we have everything separated by components it's messier to have fancy naming conventions.

kipply commented on src/components/Carousel/Carousel.jsx:
```diff
@@ -0,0 +1,55 @@
+                        />
+                    </div>
+                    <button className={styles['carousel__eye-button']} style={{backgroundColor: '#5CD568'}}/>
+                    <button className={styles['carousel__download-button']} style={{backgroundColor: '#33C9AD'}}/>
```
adhere to colours specified in https://user-images.githubusercontent.com/64981786/82742743-d8877080-9d2f-11ea-9a88-5a6740b22ae9.png

thinking-tower replied:
At the moment I'm not using the correct colours because I think we're using images?

kipply replied:
the carousel images are images, but the buttons are buttons and thus should adhere to the colour scheme? I'm confused what you're saying

kipply commented on src/components/Carousel/Carousel.jsx:
```diff
@@ -0,0 +1,55 @@
+            </div>
+            <div className={styles['carousel__project-text']}>
+                <p className={styles['carousel__project-text-title']}>
+                    ABOUT PROJECT
```
use CSS to style text to uppercase, better for SEO!

kipply commented on src/components/Carousel/Carousel.jsx:
```diff
@@ -0,0 +1,55 @@
+            <div className={styles['carousel__buttons']}>
+                <div className={styles['carousel__buttons-container']}>
+                    <div className={styles['carousel__direction-buttons-container']}>
+                        <button style={{backgroundColor: 'blue'}} 
```
adhere to https://user-images.githubusercontent.com/64981786/82742743-d8877080-9d2f-11ea-9a88-5a6740b22ae9.png

kipply commented on src/components/Carousel/Carousel.module.css:
```diff
@@ -0,0 +1,66 @@
+
+.carousel__buttons-container button {
+    border-color: transparent;
+    height: 50px;
```
smaller buttons may be a good idea

thinking-tower replied:
How much smaller?

kipply replied:
pretty small, check out some examples with nice ux; 

https://getbootstrap.com/docs/4.4/components/carousel/

kipply commented on src/components/Carousel/Carousel.module.css:
```diff
@@ -0,0 +1,66 @@
+    border-color: transparent;
+    height: 50px;
+    width: 50px;
+    border-radius: 25px;
```
you can use "50%" to make a circle instead of hardcoding half the width

kipply commented on src/components/Carousel/CarouselTrack.jsx:
```diff
@@ -0,0 +1,62 @@
+            if (index-i >= 0) {
+                distancesFromCenter[index-i] = distancesFromCenter[index-i+1]-1;
+            }
+            if (index+i < childrenWithKeys.length) {
```
else if to make it easier to follow through logic

kipply commented on src/components/Carousel/CarouselTrack.jsx:
```diff
@@ -0,0 +1,62 @@
+        return distancesFromCenter;
+    }, [maxSideLength, childrenWithKeys, index])
+
+    const style = {
```
where is this used?

kipply commented on src/components/Carousel/CarouselItem.jsx:
```diff
@@ -0,0 +1,33 @@
+    return (
+        <div className={styles['carousel__item']} style={style}>
+            {React.cloneElement(children, {
+                style: {
```
we have stylesheets, so let's avoid inline styles

thinking-tower committed:
Changes:
	- Use absolute positioning instead of marginRight and transform
	for carousel slide positioning.
	- Use images from unsplashed.com.
	- Remove font import.
	- Add example to front page.

thinking-tower committed:
Changes:
	- Use absolute positioning instead of marginRight and transform
	for carousel slide positioning.
	- Use images from unsplashed.com.
	- Remove font import.
	- Add example to front page.

thinking-tower committed:
Remove white space below.

thinking-tower committed:
Remove white space below.

thinking-tower committed:
Remove children class.

thinking-tower committed:
Remove children class.

thinking-tower committed:
Add padding to cover white space. Change font to resolve merge conflict.

thinking-tower committed:
Add padding to cover white space. Change font to resolve merge conflict.

thinking-tower commented:
## Preview
### Desktop
<img width="820" alt="Screen Shot 2020-06-09 at 11 26 25 PM" src="https://user-images.githubusercontent.com/37529096/84142337-b66c4d00-aaa8-11ea-9ecd-3f972559a083.png">

### Mobile
<img width="349" alt="Screen Shot 2020-06-09 at 11 26 45 PM" src="https://user-images.githubusercontent.com/37529096/84142356-bcfac480-aaa8-11ea-9db9-b88596cd4f90.png">

### TODOs
- Get assets for buttons
- Project text should reflect whichever template is the current slide

## References
- [metafizzy/flickity](https://github.com/metafizzy/flickity)
- [useWindowSize](https://usehooks.com/useWindowSize/)

thinking-tower committed:
Fix linting.

thinking-tower committed:
Fix merge conflict.

thinking-tower committed:
Remove unused code. Add missing dependencies for useMemo.

thinking-tower committed:
Remove unused code.
